### PR TITLE
Enable strict null checks

### PR DESCRIPTION
**Pull request by novemberborn**
Enables strict null checks and fixes violations. Some changes impact exported typings but I don't think it breaks behavior for consuming libraries that have not yet enabled strict null checks.

Depends on <https://github.com/dojo/shim/pull/19> and <https://github.com/dojo/core/pull/194> being on npm.

### CODE REVIEW

**Comment by novemberborn** [comment]
Done!

**Comment by novemberborn** [comment]
Pushed a commit which specifies the fixed `dojo-core` and `dojo-shim` versions. Somewhere in reinstalling peer dependencies a new type error cropped up (see last commit). Not sure where that came from…

File: src/mixins/createFormFieldMixin.ts [hunk 1]
```diff
@@ -77,66 +71,73 @@ export interface FormField<V> {
...
+
+					if (this.type) {
+						props['type'] = this.type;
+					}
+					/* value should always be copied */
+					props.value = this.value;
+					if (this.state && this.state.name) {
+						props.name = this.state.name;
+					}
+					if (this.state.disabled) {
```
Comment (agubler) [inline src/mixins/createFormFieldMixin.ts]: @novemberborn this has already been fixed and merged into master.... if you revert to before this commit and pull the latest then it "should" work as expected.

**Comment by novemberborn** [comment]
Rebased, bumped Maquette in the dependencies, fixed up an existing undefined guard and silenced a whole bunch more errors in the tests.

File: src/createResizePanel.ts [hunk 2]
```diff
@@ -50,23 +50,27 @@ function setResizeListeners(resizePanel: ResizePanel): Handle {
...
 
 	function onmouseupListener(evt: MouseEvent): boolean {
-		if (resizingMap.get(resizePanel)) {
-			evt.preventDefault();
-			resizingMap.delete(resizePanel);
-			onmousemoveHandle.destroy();
-			onmouseupHandle.destroy();
-			resizePanel.invalidate();
-			return true;
+		if (!resizingMap.get(resizePanel)) {
```
Comment (kitsonk) [inline src/createResizePanel.ts]: Because I am an idiot...

File: src/mixins/createParentListMixin.ts [hunk 3]
```diff
@@ -52,17 +52,20 @@ export interface ParentListMixinFactory extends ComposeFactory<ParentListMixin<C
...
+		set children(this: ParentListMixin<Child>, value: List<Child>) {
 			if (!value.equals(childrenMap.get(this))) {
 				value.forEach((widget) => {
-					if (widget.parent !== this) {
-						widget.parent = this;
-						/* TODO: If a child gets attached and reattached it may own multiple handles */
-						getRemoveHandle(this, widget);
+					// Workaround for https://github.com/facebook/immutable-js/pull/919
+					if (widget) {
+						if (widget.parent !== this) {
```
Comment (kitsonk) [inline src/mixins/createParentListMixin.ts]: That should work, right?

File: src/mixins/createParentMapMixin.ts [hunk 4]
```diff
@@ -80,10 +80,13 @@ const createParentMapMixin: ParentMapMixinFactory = compose<ParentMap<Child>, Pa
 		set children(this: ParentMapMixin<Child> & { invalidate?(): void; }, value: Map<string, Child>) {
 			if (!value.equals(childrenMap.get(this))) {
 				value.forEach((widget) => {
-					if (widget.parent !== this) {
-						widget.parent = this;
-						/* TODO: If a child gets attached and reattached it may own multiple handles */
-						getRemoveHandle(this, widget);
+					// Workaround for https://github.com/facebook/immutable-js/pull/919
+					if (widget) {
```
Comment (kitsonk) [inline src/mixins/createParentMapMixin.ts]: Again, `widget & widget.parent !== this` doesn't work?

File: src/createResizePanel.ts (hunk 2 above)
Comment (novemberborn) [inline src/createResizePanel.ts]: Heh I can change that. I focused on the errors, not the code itself.

File: src/mixins/createParentListMixin.ts (hunk 3 above)
Comment (novemberborn) [inline src/mixins/createParentListMixin.ts]: Yes that would work but it obscures the Immutable workaround.

File: src/mixins/createTabbedMixin.ts [hunk 5]
```diff
@@ -84,7 +87,13 @@ function setActiveTab(tabbed: TabbedMixin<TabbedChild>, activeTab: TabbedChild)
  * @param tabbed The tabbed mixin to return the active child for
  */
 function getActiveTab(tabbed: TabbedMixin<TabbedChild>): TabbedChild {
-	let activeTab = tabbed.children.find((tab) => tab.state.active);
+	let activeTab = tabbed.children.find((tab) => {
+		// Workaround for https://github.com/facebook/immutable-js/pull/919
+		if (!tab) {
+			return false;
```
Comment (kitsonk) [inline src/mixins/createTabbedMixin.ts]: This won't ever get covered, so istanbul ignore next or:

```typescript
let activeTab = tabbed.children.find((tab) => tab && tab.state.active);
```

File: src/mixins/createVNodeEvented.ts [hunk 6]
```diff
@@ -137,10 +137,12 @@ function handlesArraytoHandle(handles: Handle[]): Handle {
 	};
 }
 
+const UNINITIALIZED_LISTENERS = Object.freeze({});
```
Comment (kitsonk) [inline src/mixins/createVNodeEvented.ts]: `Object.freeze(Object.create(null))` or is that just silly?

File: src/mixins/createVNodeEvented.ts [hunk 7]
```diff
@@ -137,10 +137,12 @@ function handlesArraytoHandle(handles: Handle[]): Handle {
...
+
 const createVNodeEvented: VNodeEventedFactory = createEvented
 	.mixin({
 		mixin: {
-			listeners: <VNodeListeners> null,
+			listeners: UNINITIALIZED_LISTENERS,
```
Comment (kitsonk) [inline src/mixins/createVNodeEvented.ts]: 👍 awesomeness

File: src/projector.ts [hunk 8]
```diff
@@ -180,18 +180,25 @@ export const createProjector: ProjectorFactory = compose<ProjectorMixin, Project
...
+		attach(this: Projector, { type, tagName }: AttachOptions = {}): Promise<Handle> {
 			const projectorData = projectorDataMap.get(this);
 			if (projectorData.state === ProjectorState.Attached) {
-				return projectorData.attachPromise;
+				return projectorData.attachPromise || Promise.resolve(noopHandle);
 			}
 			projectorData.boundRender = this.render.bind(this);
-			projectorData.tagName = tagName;
+			if (tagName !== undefined) {
+				projectorData.tagName = tagName;
```
Comment (kitsonk) [inline src/projector.ts]: this is not getting covered, but the surrounding code was...

File: src/projector.ts [hunk 9]
```diff
@@ -326,6 +336,42 @@ export const createProjector: ProjectorFactory = compose<ProjectorMixin, Project
 		}
 	});
 
-const defaultProjector: Projector = typeof global.document === 'undefined' ? null : createProjector();
+// Projectors cannot be created outside of browser environments. Ensure that a default projector can always be
+// exported, even if it can't do anything.
+const createStubbedProjector: ProjectorFactory = compose({
+		getNodeAttributes(): VNodeProperties {
+			throw new Error('Projector is stubbed');
```
Comment (kitsonk) [inline src/projector.ts]: This is new code that was generally being covered but now isn't.

File: src/mixins/createParentMapMixin.ts (hunk 4 above)
Comment (novemberborn) [inline src/mixins/createParentMapMixin.ts]: It works, but I've tried to keep the workaround separate from the original code.

File: src/mixins/createVNodeEvented.ts (hunk 6 above)
Comment (novemberborn) [inline src/mixins/createVNodeEvented.ts]: It gets replaced anyhow, so seems unnecessary.

File: src/mixins/createParentListMixin.ts (hunk 3 above)
Comment (kitsonk) [inline src/mixins/createParentListMixin.ts]: Ok, I am fine leaving it then.

File: src/projector.ts (hunk 8 above)
Comment (novemberborn) [inline src/projector.ts]: There are no tests that pass options to `attach()`, so previously when we assigned the default while destructuring the argument this line would get covered by the existing test. This is now uncovered like a bunch of other stuff in the function that relies on options being passed.

File: src/projector.ts (hunk 9 above)
Comment (novemberborn) [inline src/projector.ts]: > Will ignore from code coverage.

Actually no, I won't. But I can't really fix it yet either.

**Comment by novemberborn** [comment]
Pushed fixups for the remaining stuff.

**Pull request merged**